---
title: 'jDMRgrid: a heuristic DMR caller for WGBS data using grid approach'
author: "Robert S. Piecyk, Rashmi R. Hazarika, Yadi Shahryary, Frank Johannes"
date: "2024-10-31"
output:
  pdf_document:
    toc: true
    toc_depth: 2
  html_document:
    toc: true
    toc_depth: 2
---

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document.

```{r}
setwd("/jlab/home/ming/rmarkdown")
```


## 1 Input files

For generation of region-level calls, jDMRgrid requires the following inputs.

## 1.1 Methimpute files:

Base-level methylome outputs (generated using the R package "Methimpute")

## 1.2 A metadata file containing description about samples

For population data-sets without replicates, listfiles.fn should have the structure below.

```{r}
load(system.file("data", "listFiles1.RData", package = "jDMRgrid"))
listFiles1$file <- system.file("extdata", listFiles1$file, package = "jDMRgrid")
listFiles1
```

For pairwise control-treatment data-sets with replicates,additional columns "replicate" and "group" should be provided. See
structure below.

```{r}
load(system.file("data", "listFiles2.RData", package = "jDMRgrid"))
listFiles2$file <- system.file("extdata", listFiles2$file, package = "jDMRgrid")
listFiles2
```

\newpage

## 2 Generate cytosine region calls from genome

## 2.1 Run jDMRgrid on a binned genome

This function uses a grid approach to bin the genome into equal sized bins. User specifies the window and step size as numeric values.

**out.dir:** PATH to output directory.

**window:** NUMERIC VALUE specifying bin size.

**step:** NUMERIC VALUE specifying step size. If bin and step size are equal, we are utilizing non-sliding window approach.

**samplelist:** DATAFRAME OBJECT containing information about file, sample and replicate. For control/treatment data an additional column specifying the replicates is required.

**contexts:** VECTOR or CHARACTER presenting sequence contexts of the cytosine. By default this option is set to
c("CG", "CHG", "CHH", "CHH"). If you want to run for a single context such as CG, set it as "CG".

**min.C:** NUMERIC VALUE specifying percentile threshold based on empirical distribution of the cytosines across bins.

**mincov:** NUMERIC VALUE specifying minimum read coverage over cytosines. By default this option is set as 0.

**include.intermediate:** LOGICAL specifying whether or not the intermediate component should be included in the HMM model. By default this option is set as FALSE.

**runName:** CHARACTER as the name of the operation. By default this option is set to ‘GridGenome’.

**parallelApply:** LOGICAL specifying if future.apply package should be used to use parallel operation. By default this
option is set to FALSE.

**numCores:** NUMERIC VALUE specifying number of cores to perform parallel operation using foreach loop. By default this option is set to NULL.

**if.Bismark:** Logical if Bismark inputs (CX reports in txt format) are used. Default as FALSE. (logical)

**FASTA.file:** Path to the FASTA file; required if Bismark outputs are used. Please use the same FASTA as in Bismark analysis. Default as NULL. (char)

```{r echo = TRUE, message = FALSE, warning=FALSE, results='hide'}
library(jDMRgrid)
runjDMRgrid(out.dir = "./population",
            window = 200, step = 50, samplelist = listFiles1,
            contexts = c("CG", "CHG", "CHH"), min.C = 10, mincov = 0,
            include.intermediate = TRUE,
            runName = "Arabidopsis")

runjDMRgrid(out.dir = "./replication",
            window = 200, step = 50, samplelist = listFiles2,
            contexts = c("CG", "CHG", "CHH"), min.C = 10, mincov = 0,
            include.intermediate = TRUE,
            runName = "Arabidopsis")
```

## 2.2 Output files of jDMR Grid approach

Region files containing state calls, methylation levels and posteriorMax will be generated for each sample and for each context.

```{r, message=FALSE}
library(data.table)
jDMRgrid.out <- fread("./population/methimpute_p1_CG.txt", header = T)
head(jDMRgrid.out)
```

**seqnames, start and end:** Chromosome coordinates

**context:** Sequence context of cytosine i.e CG,CHG,CHH

**posteriorMax:** Posterior value of the methylation state call

**status:** Methylation status

**rc.meth.lvl:** Recalibrated methylation level calculated from the posteriors and fitted parameters

\newpage

## 3 Generate DMR matrix

## 3.1 Run "makeDMRmatrix"

This function generates a DMR matrix of state calls, rc.meth.lvls and posterior probabilities for all samples in one dataframe.

**samplelist:** DATAFRAME OBJECT containing information about file, sample and replicate. For control/treatment data an additional column specifying the replicates is required.

**input.dir:** PATH to directory containing region files.

**out.dir:** PATH to output directory.

**contexts:** sequence contexts of the cytosine. By default this option is set to c("CG", "CHG", "CHH", "CHH"). If you want to run for a single context such as CG, set it as "CG".

**postMax.out:** By default this option is set as FALSE. You can set it to TRUE if you want to output the DMR matrix containing posterior probabilities for the status call of each region.

**include.intermediate:** LOGICAL specifying whether or not the intermediate component should be included in the HMM model. By default this option is set as FALSE.

```{r, message=FALSE}
makeDMRmatrix(contexts = c("CG", "CHG", "CHH"), postMax.out = TRUE,
              samplelist = listFiles1,
              input.dir = "./population",
              out.dir = "./population/matrix",
              include.intermediate = TRUE)

makeDMRmatrix(contexts = c("CG", "CHG", "CHH"), postMax.out = TRUE,
              samplelist = listFiles2,
              input.dir = "./replication",
              out.dir = "./replication/matrix",
              include.intermediate = TRUE)
```

## 3.2 Output files of makeDMRmatrix function

"CG_StateCalls.txt" has the following structure. "0" in the output matrix denotes "Unmethylated", "1" stands for "Methylated" and "0.5" stands for "Intermediate".

```{r}
statecalls <- fread("./population/matrix/CG_StateCalls.txt", header = T)
head(statecalls)
```

"CG_rcMethlvl.txt" has the following structure. The output matrix contains recalibrated methylation levels for each sample and for the specific region.

```{r}
rcmethlvls <- fread("./population/matrix/CG_rcMethlvl.txt", header = T)
head(rcmethlvls)
```

"CG_postMax.txt" has the following structure. The output matrix contains posterior probabilities for each sample and for the specific region.

```{r}
postMax <- fread("./population/matrix/CG_postMax.txt", header = T)
head(postMax)
```

## 3.3 Split DMR matrix into pairwise groups

Ignore this step if you are running jDMR on population data without replicates. For pairwise control-treatment data-sets with replicates, you need to run this step. This function generates a DMR matrix of state calls, rc.meth.lvls and posterior probabilities for each pairwise control-treatment in one dataframe. For example, if the samplelist includes control, treatment1 and treatment2, this function generates a DMR matrix of state calls, rc.meth.lvls and posterior probabilities for control-treatment1 in one dataframe, and a DMR matrix of state calls, rc.meth.lvls and posterior probabilities for control-treatment2 in one dataframe.

**samplelist:** DATAFRAME OBJECT containing information about file, sample and replicate. For control/treatment data an additional column specifying the replicates is required.

**input.dir:** PATH to directory containing region files.

**out.dir:** PATH to output directory.

**contexts:** sequence contexts of the cytosine. By default this option is set to c("CG", "CHG", "CHH"). If you want to run for a single context such as CG, set it as "CG".

**postMax.out:** by default this option is set to FALSE. If you want to output the matrix containing posterior probabilities set it to TRUE.

```{r, message=FALSE}
splitGroups(samplelist = listFiles2, postMax.out = TRUE, contexts = c("CG", "CHG", "CHH"),
            input.dir = "./replication/matrix",
            out.dir = "./replication/matrix")
```

## 3.4 Output files of splitGroups function

```{r}
statecalls_split <- fread("./replication/matrix/CG_WT_mutant1_StateCalls.txt", header = T)
head(statecalls_split)
```

\newpage

## 4 Filter DMR matrix

## 4.1 Filter the DMR matrix

This function filters the DMR matrix for non-polymorphic patterns.

For population data-sets without replicates, this function will find the outputs of 'runjDMRgrid' for filtering. For pairwise control-treatment data-sets with replicates, this function will find the outputs of 'splitGroups' for filtering.

**data.dir:** PATH to folder containing DMR matrix

**epiMAF.cutoff:** Applicable for calling calling population DMRs. This option can be used to filter for Minor Epi-Allele frequency as specified by user. By default, this option is set to NULL.

**replicate.consensus:** Applicable for control-treatment data-sets with replicates. Users can specify the percentage of concordance in methylation states in samples with multiple replicates. For datasets with just 2 replicates, replicate.consensus should be set as 1 (means 100% concordance). By default, this option is set to NULL.

**samplelist:** DATAFRAME OBJECT containing information about file, sample and replicate. For control/treatment data an additional column specifying the replicates is required.

**if.mergingBins:** Logical argument if merging consecutive bins having the same stateCalls should be performed. By default set to TRUE. (logical)

```{r echo = TRUE, message = FALSE, warning=FALSE, results='hide'}
filterDMRmatrix(epiMAF.cutoff = NULL, replicate.consensus = 0.5,
                data.dir = "./population/matrix",
                samplelist = listFiles1, if.mergingBins = FALSE)
```

```{r echo = TRUE, message = FALSE, warning=FALSE, results='hide'}
filterDMRmatrix(epiMAF.cutoff = NULL, replicate.consensus = 0.5,
                data.dir = "./replication/matrix",
                samplelist = listFiles2, if.mergingBins = FALSE)
```

## 4.2 Filtered Output

"CG_StateCalls-filtered.txt" has the following structure.

```{r}
statecalls_filtered <- fread("./population/matrix/CG_StateCalls-filtered.txt", header = T)
head(statecalls_filtered)

statecalls_filtered <- fread("./replication/matrix/CG_WT_mutant1_StateCalls-filtered.txt", header = T)
head(statecalls_filtered)
```

\newpage

## 5 Search for context-specific and annotate DMRs

## 5.1 Output context specific DMR

Output DMRs specific for contexts i.e CG-only, CHG-only, CHH-only, non-CG and multi-context DMRs using the StateCalls_filtered.txt files (if variable if Filtered equals to TRUE) or StateCalls.txt files (if variable ifFiltered equals to FALSE, as default).

```{r, message=FALSE}
context.specific.DMRs(samplelist = listFiles2,
                      output.dir="./replication/context_DMRs",
                      input.dir = "./replication/matrix",
                      if.filtered = TRUE)
```

## 5.2 Annotate DMRs

This function annotates the lists of DMRs. Any file(.txt) containing 3 columns (chr, start, end) can be annotated using the annotateDMRs function. Please move all files to be annotated to a separate folder and set the full PATH to the "input.dir" option.

**gff.files:** Multiple gff3 annotation files can be supplied as a vector

**annotation:** specify annotation categories

**input.dir:** path to folder containing only files to be annotated. Any file containing 3 columns (chr, start, end) can be annotated using the annotateDMRs function.

**if.gff3:** whether to output annotated files in gff3 format

**out.dir:** path to output folder

In the following example, I will annotate the files generated in section 3.1

```{r, message=FALSE}
gff.file_promoters <- system.file("extdata/toyData", "TAIR10_promoters.gff3", package = "jDMRgrid")
gff.file_TE <- system.file("extdata/toyData", "TAIR10_TE.gff3", package = "jDMRgrid")

annotateDMRs(gff.files = c(gff.file_promoters, gff.file_TE),
             annotation = c("promoters", "TE"),
             input.dir = "./replication/annotation",
             if.gff3 = FALSE,
             out.dir = "./replication/annotation")
```

## 5.3 Output files after annotation

Mapped files are output in .txt and/or .gff3 format. Addiitonally, a DMR count table is generated.

```{r, message=FALSE}
annotations <- fread("./replication/annotation/CG_WT_mutant1_StateCalls_annotation.txt",
                     header = T)
head(annotations)

DMRcounts <- fread("./replication/annotation/DMR-counts.txt", header = T)
head(DMRcounts)
```

\newpage

## 6 R session info

```{r}
sessionInfo()
```

