---
title: "jDMRgrid: a heuristic DMR caller for WGBS data using grid approach"
author: 'Robert S. Piecyk, Rashmi R. Hazarika, Yadi Shahryary, Frank Johannes '
date: "`r Sys.Date()`"
output:
    pdf_document:
        fig_caption: yes
        number_sections: yes
        toc: yes
        toc_depth: 4
    word_document:
        toc: yes
        toc_depth: '4'
    html_document:
        df_print: paged
        toc: yes
        toc_depth: '4'
geometry: margin=0.5in
urlcolor: blue
vignette: |
    %\VignetteIndexEntry{jDMRgrid} 
    %\VignetteEncoding{UTF-8}
    %\usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::knitr} 
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include = FALSE}
options(width=120)
knitr::opts_chunk$set(
    collapse = FALSE,
    eval = TRUE,
    comment = " ",
    tidy.opts =list(width.cutoff=80),
    tidy = TRUE,
    size="small"
)
```

\newpage

```{r eval=TRUE, include=FALSE}
library(jDMRgrid)
library(data.table)
```

\section{Input files}

For generation of region-level calls, jDMRgrid requires the following
inputs.

## Methimpute files:

Base-level methylome outputs (generated using the R package
"Methimpute")

## A metadata file containing description about samples

For population data-sets without replicates, listfiles.fn should have
the structure below.

\setlength{\leftskip}{1cm}

**file**: full PATH of file.
**sample**: a sample name

\setlength{\leftskip}{0pt}

```{r, include=TRUE}
load(system.file("data", "listFiles1.RData", package="jDMRgrid"))
listFiles1$file <- system.file("extdata", listFiles1$file,  package="jDMRgrid")
listFiles1
```

For pairwise control-treatment data-sets with replicates,additional
columns "replicate" and "group" should be provided. See structure below.

\setlength{\leftskip}{1cm}

**file**: full PATH of file
**sample**: a sample name
**replicate**: label for replicates
**group**: label for control and treatment groups

\setlength{\leftskip}{0pt}

```{r, include=TRUE}
load(system.file("data", "listFiles2.RData", package="jDMRgrid"))
listFiles2$file <- system.file("extdata", listFiles2$file,  package="jDMRgrid")
listFiles2
```

```{r eval=TRUE, include=FALSE}
system2(command = "/bin/mkdir", args = "~/folder_population")
system2(command = "/bin/mkdir", args = "~/folder_replicate")
system2(command = "/bin/mkdir", args = "~/folder_population/grid")
system2(command = "/bin/mkdir", args = "~/folder_replicate/grid")
system2(command = "/bin/mkdir", args = "~/folder_population/matrix")
system2(command = "/bin/mkdir", args = "~/folder_replicate/matrix")
system2(command = "/bin/mkdir", args = "~/folder_population/context_DMRs")
system2(command = "/bin/mkdir", args = "~/folder_replicate/context_DMRs")
system2(command = "/bin/mkdir", args = "~/folder_population/annotate_DMRs")
system2(command = "/bin/mkdir", args = "~/folder_replicate/annotate_DMRs")
```

```{=tex}
\newpage
\section{Generate cytosine region calls from genome}
```
jDMR detects DMRs using two approaches a) finding cytosine clusters in
the genome (section 2.1) b) using a binning approach (section 2.2). You
can use either of the methods to obtain the region calls. The remaining
steps, makeDMRmatrix, filterDMRmatrix, annotateDMRs are the same for
both methods.

## Run jDMRgrid on a binned genome

This function uses a grid approach to bin the genome into equal sized
bins. User specifies the window and step size as numeric values.

\setlength{\leftskip}{1cm}

**out.dir**: PATH to output directory.

**window**: NUMERIC VALUE specifying bin size.

**step**: NUMERIC VALUE specifying step size. If bin and step size are
equal, we are utilizing non-sliding window approach.

**samplelist**: DATAFRAME OBJECT containing information about file,
sample and replicate. For control/treatment data an additional column
specifying the replicates is required.

**contexts**: VECTOR or CHARACTER presenting sequence contexts of the
cytosine. By default this option is set to c("CG", "CHG", "CHH"). If you
want to run for a single context such as CG, set it as "CG".

**min.C**: NUMERIC VALUE specifying percentile threshold based on
empirical distribution of the cytosines across bins.

**mincov**: NUMERIC VALUE specifying minimum read coverage over
cytosines. By default this option is set as 0.

**include.intermediate**: LOGICAL specifying whether or not the
intermediate component should be included in the HMM model. By default
this option is set as FALSE.

**runName**: CHARACTER as the name of the operation. By default this
option is set to 'GridGenome'.

**parallelApply**: LOGICAL specifying if future.apply package should be
used to use parallel operation. By default this option is set to FALSE.

**numCores**: NUMERIC VALUE specifying number of cores to perform
parallel operation using foreach loop. By default this option is set to
NULL.

\setlength{\leftskip}{0pt}

```{r eval=FALSE, include=TRUE}
library(jDMRgrid)
runjDMRgrid(
    out.dir = '~/folder_population/grid',
    window = 200, step = 50, samplelist = listFiles1,
    contexts = c("CG", "CHG", "CHH"), min.C = 10, mincov = 0,
    include.intermediate = TRUE, runName = "Arabidopsis")
runjDMRgrid(
    out.dir = '~/folder_replicate/grid',
    window = 200, step = 50, samplelist = listFiles2,
    contexts = c("CG", "CHG", "CHH"), min.C = 10, mincov = 0,
    include.intermediate = TRUE, runName = "Arabidopsis")
```

### Output files of jDMR Grid approach

Region files containing state calls and methylation levels will be
generated for each sample and for each context.

```{r eval=TRUE, include=TRUE}
jDMR.out <- fread('~/folder_replicate/grid/methimpute_p1_CG.txt')
```

```{r echo=FALSE}
head(jDMR.out)
```

\setlength{\leftskip}{1cm}

**seqnames, start and end**: Chromosome coordinates

**context**: Sequence context of cytosine i.e CG,CHG,CHH

**posteriorMax**: Posterior value of the methylation state call

**status** : Methylation status

**rc.meth.lvl**: Recalibrated methylation level calculated from the
posteriors and fitted parameters

\setlength{\leftskip}{0pt}

```{=tex}
\newpage
\section{Generate DMR matrix}
```
## Run "makeDMRmatrix"

This function generates a DMR matrix of state calls, rc.meth.lvls and
posterior probabilities for all samples in one dataframe.

\setlength{\leftskip}{1cm}

**samplelist**: DATAFRAME OBJECT containing information about file,
sample and replicate. For control/treatment data an additional column
specifying the replicates is required.

**input.dir**: PATH to directory containing region files.

**out.dir**: PATH to output directory.

**contexts**: sequence contexts of the cytosine. By default this option
is set to c("CG", "CHG", "CHH"). If you want to run for a single context
such as CG, set it as "CG".

**postMax.out**: By default this option is set as FALSE. You can set it
to TRUE if you want to output the DMR matrix containing posterior
probabilities for the status call of each region.

\setlength{\leftskip}{0pt}

```{r eval=FALSE, include=TRUE}
makeDMRmatrix(
    contexts = c("CG", "CHG", "CHH"), postMax.out = TRUE, 
    samplelist = listFiles1,
    input.dir = '~/folder_population/grid', 
    out.dir = '~/folder_population/matrix', include.intermediate = FALSE)

makeDMRmatrix(
    contexts = c("CG", "CHG", "CHH"), postMax.out = TRUE,
    samplelist = listFiles2,
    input.dir = '~/folder_replicate/grid',
    out.dir = '~/folder_replicate/matrix', include.intermediate = FALSE)
```

## Output files of DMRmatrix function

*"CG_StateCalls.txt" has the following structure. "0" in the output
matrix* *denotes "Unmethylated" and "1" stands for "Methylated".*

```{r eval=TRUE, include=TRUE}
statecalls <- fread("~/folder_replicate/matrix/CG_StateCalls.txt", header=TRUE)
```

```{r echo=FALSE}
head(statecalls)
```

*"CG_rcMethlvl.txt" has the following structure. The output matrix
contains* *recalibrated methylation levels for each sample and for the
specific region.*

```{r eval=TRUE, include=TRUE}
rcmethlvls <- fread("~/folder_replicate/matrix/CG_rcMethlvl.txt", header=TRUE)
```

```{r echo=FALSE}
head(rcmethlvls)
```

*"CG_postMax.txt" has the following structure. The output matrix
contains* *posterior probabilities for each sample and for the specific
region.*

```{r eval=TRUE, include=TRUE}
postMax <- fread("~/folder_replicate/matrix/CG_postMax.txt", header=TRUE)
```

```{r echo=FALSE}
head(postMax)
```

## Split DMR matrix into pairwise groups

Ignore this step if you are running jDMR on population data without
replicates

\setlength{\leftskip}{1cm}

**samplelist**: DATAFRAME OBJECT containing information about file,
sample and replicate. For control/treatment data an additional column
specifying the replicates is required.

**input.dir**: PATH to directory containing region files.

**out.dir**: PATH to output directory.

**contexts**: sequence contexts of the cytosine. By default this option
is set to c("CG", "CHG", "CHH"). If you want to run for a single context
such as CG, set it as "CG".

**postMax.out**: by default this option is set to FALSE. If you want to
output the matrix containing posterior probabilities set it to TRUE.

\setlength{\leftskip}{0pt}

```{r eval=FALSE, include=TRUE}
splitGroups(
    samplelist = listFiles2,
    input.dir = '~/folder_replicate/matrix',
    out.dir = '~/folder_replicate/matrix')
```

```{=tex}
\newpage
\section{Filter DMR matrix}
```
## Filter the DMR matrix

This function filters the DMR matrix for non-polymorphic patterns.

\setlength{\leftskip}{1cm}

**data.dir**: PATH to folder containing DMR matrix

**epiMAF.cutoff**: Applicable for calling calling population DMRs. This
option can be used to filter for Minor Epi-Allele frequency as specified
by user. By default, this option is set to NULL.

**replicate.consensus** : Applicable for control-treatment data-sets
with replicates. Users can specify the percentage of concordance in
methylation states in samples with multiple replicates. For datasets
with just 2 replicates, \textit{replicate.consensus} should be set as 1
(means 100% concordance). By default, this option is set to NULL.

**samplelist**: DATAFRAME OBJECT containing information about file,
sample and replicate. For control/treatment data an additional column
specifying the replicates is required.

**if.mergingBins** : Logical argument if merging consecutive bins having
the same stateCalls should be performed. By default set to TRUE.
(logical)

\setlength{\leftskip}{0pt}

```{r eval=FALSE, include=TRUE, echo=FALSE}
filterDMRmatrix(
    epiMAF.cutoff = 0.33, replicate.consensus = NULL, 
    data.dir = "~/folder_population/matrix", 
    samplelist = listFiles1, if.mergingBins = FALSE)
filterDMRmatrix(
    epiMAF.cutoff = NULL, replicate.consensus = 0.8, 
    data.dir = "~/folder_replicate/matrix", 
    samplelist = listFiles2, if.mergingBins = FALSE)
```

## Filtered Output

*"CG_StateCalls-filtered.txt" has the following structure.*

```{r eval=TRUE, include=TRUE}
statecallsFiltered <- fread(
    "~/folder_population/matrix/CG_StateCalls-filtered.txt", header=TRUE)
```

```{r echo=FALSE}
head(statecallsFiltered)
```
If "rc.methlvl.out" option is set to TRUE a filtered matrix with
averaged methylation levels in generated.

```{r eval=TRUE, include=TRUE}
rcmethlvlFiltered <- fread(
    "~/folder_population/matrix/CG_rcMethlvl-filtered.txt", header=TRUE)
```

```{r echo=FALSE}
head(rcmethlvlFiltered)
```

```{=tex}
\newpage
\section{Search for context-specific and annotate DMRs}
```
## Output context specific DMRs

Output DMRs specific for contexts i.e CG-only, CHG-only, CHH-only,
non-CG and multi-context DMRs using the *StateCalls-filtered.txt files
(if variable ifFiltered equals to TRUE) or* StateCalls.txt files (if
variable ifFiltered equals to FALSE, as default).

```{r eval=FALSE, include=TRUE}
context.specific.DMRs(
    samplelist = listFiles1,
    output.dir = "~/folder_population/context_DMRs",
    input.dir = "~/folder_population/matrix", if.filtered = FALSE)
context.specific.DMRs(
    samplelist = listFiles2,
    output.dir = "~/folder_replicate/context_DMRs",
    input.dir = "~/folder_replicate/matrix", if.filtered = FALSE)
```

## Annotate DMRs

This function annotates the lists of DMRs. Any file(.txt) containing 3
columns (chr, start, stop) can be annotated using the annotateDMRs
function. Please move all files to be annotated to a separate folder and
set the full PATH to the "input.dir" option.

\setlength{\leftskip}{1cm}

**gff.files**: Multiple gff3 annotation files can be supplied as a
vector

**annotation**: specify annotation categories

**input.dir**: path to folder containing only files to be annotated. Any
file containing 3 columns (chr, start, stop) can be annotated using the
annotateDMRs function.

**if.gff3**: whether to output annotated files in gff3 format

**out.dir**: path to output folder

\setlength{\leftskip}{0pt}

In the following example, I will annotate the files generated in section
4.3

```{r eval=FALSE, include=TRUE}
gff.file_promoters <-system.file(
    "extdata/toyData","TAIR10_promoters.gff3", package="jDMRgrid")
gff.file_TE <- system.file(
    "extdata/toyData","TAIR10_TE.gff3", package="jDMRgrid")

annotateDMRs(
    gff.files = c(gff.file_promoters, gff.file_TE),
    annotation = c("promoters", "TE"),
    input.dir = "~/folder_population/context_DMRs", if.gff3 = FALSE,
    out.dir = '~/folder_population/annotate_DMRs')

annotateDMRs(
    gff.files = c(gff.file_promoters, gff.file_TE),
    annotation = c("promoters", "TE"),
    input.dir = "~/folder_replicate/context_DMRs", if.gff3 = FALSE,
    out.dir = '~/folder_replicate/annotate_DMRs')
```

## Output files after annotation

Mapped files are output in .txt and/or .gff3 format. Addiitonally, a DMR
count table is generated.

```{r eval=TRUE, include=TRUE}
DMRcounts <- fread(
    "~/folder_replicate/annotate_DMRs/DMR-counts.txt", header=TRUE)
```

```{r echo=FALSE}
head(DMRcounts)
```

```{=tex}
\newpage
\section{R session info }
```
```{r}
sessionInfo()
```
